### PR TITLE
prov/efa: inject related changes

### PR DESCRIPTION
**Pull request by wzamazon**
This PR fixes issue related to inject.

### CODE REVIEW

**Review by bwbarrett — CHANGES_REQUESTED** [review]

File: prov/efa/src/rxr/rxr.h [hunk 1]
```diff
@@ -567,7 +567,7 @@ struct rxr_ep {
 	/* rx/tx queue size of core provider */
 	size_t core_rx_size;
 	size_t max_outstanding_tx;
-	size_t core_inject_size;
+	size_t inject_size;
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/rxr/rxr.h]: Sure. In the new revision, I kept core_inject_size untouched.

File: prov/efa/src/rxr/rxr_init.c [hunk 2]
```diff
@@ -349,8 +350,19 @@ static int rxr_info_to_rxr(uint32_t version, const struct fi_info *core_info,
...
-	info->tx_attr->inject_size = 0;
+	/*
+	 * The requirement for inject is: upon return, the user buffer can be reused immediately.
+	 *
+	 * For EFA, inject is implement as: construct a packet entry, copy user data to packet entry
+	 * then send the packet entry. Therefore the maximum inject size is
+	 *    pkt_entry_size - maximum_header_size.
+	 */
+	if (rxr_env.enable_shm_transfer)
+		min_pkt_size = MIN(core_info->ep_attr->max_msg_size, rxr_env.shm_max_medium_size);
```
Comment (bwbarrett) [inline prov/efa/src/rxr/rxr_init.c:361]: I'm ok with this for now, but we can do better.  There's no reason that we can't copy in the EFA provider and then let shmem use the long message protocol.  That would give us a nicer transition from inject to the medium protocol for EFA itself.

File: prov/efa/src/rxr/rxr_init.c [hunk 3]
```diff
@@ -349,8 +350,19 @@ static int rxr_info_to_rxr(uint32_t version, const struct fi_info *core_info,
...
+	else
+		min_pkt_size = core_info->ep_attr->max_msg_size;
```
Comment (bwbarrett) [inline prov/efa/src/rxr/rxr_init.c:363]: max_msg_size is the size of a single EFA packet, right?

**Review by bwbarrett — APPROVED** [review]

File: prov/efa/src/rxr/rxr_init.c (hunk 2 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_init.c:361]: Sure, will implement in the future.

File: prov/efa/src/rxr/rxr_init.c (hunk 3 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_init.c:363]: Yes.

**Comment by wzamazon** [comment]
bot:aws:retest

**Comment by wzamazon** [comment]
bot:aws:retest

**Comment by wzamazon** [comment]
bot:aws:retest

**Comment by wzamazon** [comment]
bot:aws:retest

**Comment by wzamazon** [comment]
All CI tests passed, merging.

**Pull request merged**